criu/seize.c

```diff
@@ -599,6 +600,35 @@ static int freeze_processes(void)
 				goto err;
 			}
 			nanosleep(&req, NULL);
+
+			if (cgroup_v2)
+				continue;
+
+			/* As per older kernel docs (freezer-subsystem.txt before
+			 * the kernel commit ef9fe980c6fcc1821), if FREEZING is seen,
+			 * userspace should either retry or thaw. While current
+			 * kernel cgroup v1 docs no longer mention a need to retry,
+			 * even recent kernels can't reliably freeze a cgroup v1.
+			 *
+			 * Let's keep asking the kernel to freeze from time to time.
+			 * In addition, do occasional thaw/sleep/freeze.
+			 *
+			 * This is still a game of chances (the real fix belongs to the kernel)
+			 * but these kludges might improve the probability of success.
+			 *
+			 * Cgroup v2 does not have this problem.
+			 */
+			switch (i % 32) {
+			case 9:
+			case 20:
+				freezer_write_state(fd, FROZEN);
+				break;
+			case 31:
+				freezer_write_state(fd, THAWED);
```

Comment:
Kir, there is no sense to do the previous nanosleep and then thaw the cgroup.

Reply:
Right, I got it now.

Refactored to eliminate this (and some other) issues.